test/solver/solver.cpp

```diff
@@ -96,22 +98,24 @@ struct SimpleSolverTest {
 
     static typename solver_type::parameters_type build(
         std::shared_ptr<const gko::Executor> exec,
-        gko::size_type iteration_count)
+        gko::size_type iteration_count, bool check_residual = true)
     {
         return solver_type::build().with_criteria(
             gko::stop::Iteration::build()
                 .with_max_iters(iteration_count)
-                .on(exec));
+                .on(exec),
+            check_residual ? gko::stop::ResidualNorm<value_type>::build()
+                                 .with_baseline(gko::stop::mode::absolute)
+                                 .with_reduction_factor(1e-30)
```

Comment:
This is needed to prevent the residual from becoming zero, which leads to nans in some solvers. This is especially an issue for the test where the initial guess is already the solution. Without this check, Gmres, for example, will only compute nans.

Reply:
Is this a fix for the lucky breakdown then? I would assume fixing it means preventing the solver from ever underflowing the residual.

Reply:
I don't think we can prevent underflow, because in some cases, e.g. initial guess is the solution, that is the expected behavior. So we would need to make sure that we stop in that case. But I think that is difficult to realize without checking the residual twice because we don't know if a residual criterion is already in use.

Reply:
We can act as if the stopping criterion fired, setting a flag based on the implicit residual should be cheap. But maybe we should leave that for another PR.

Reply:
I briefly added a criterion to check for breakdown, but since that leads to bigger changes than expected, I moved that into another branch and will create another PR for that.